### ISSUE TITLE
Clamp gradient behaves differently in PyTorch 2.14

### ISSUE THREAD

**Issue by kyunggeunlee** [pytorch/pytorch#195931]
### Problem
Clamp gradient behaves differently in PyTorch 2.14.

Given `y = clamp(x, min, max)`,
until torch 2.13, its gradient was calculated as
```
# torch 2.13
dy/dx = 1 (if min <= x <= max)
        0 (otherwise)
```
                                                                             
However, torch 2.14 changed the gradient boundary from closed interval to open interval, i.e.
```
# torch 2.14                                                                 
dy/dx = 1 (if min < x < max)
        0 (otherwise)
```

Although technically both 2.13 and 2.14 are mathematically correct, 2.13-style gradient seems more practical. Was this an intentional change? If not, could you please consider reverting clamp gradient back to 2.13 style?

### Reproducible Example
```py
import torch


def main():
    x = torch.tensor([-2.0, -1.0, 0.0, 1.0, 2.0], requires_grad=True)
    y = torch.clamp(x, min=-1.0, max=1.0)
    y.backward(torch.ones_like(y))

    # output:
    #   - torch 2.13: "tensor([0., 1., 1., 1., 0.])"
    #   - torch 2.14: "tensor([0., 0., 1., 0., 0.])"
    print(x.grad)


if __name__ == "__main__":
    main()
```

### Versions

Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (aarch64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 4.4.3
Libc version: glibc-2.35

Python version: 3.10.16 (main, Mar 18 2025, 12:09:19) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.153.1-microsoft-standard-WSL2-aarch64-with-glibc2.35
Is CUDA available: N/A
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
Is XPU available: N/A
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A
Caching allocator config: N/A

CPU:
Architecture:                       aarch64
CPU op-mode(s):                     64-bit
Byte Order:                         Little Endian
CPU(s):                             12
On-line CPU(s) list:                0-11
Vendor ID:                          Qualcomm
Model:                              1
Thread(s) per core:                 1
Core(s) per socket:                 12
Socket(s):                          1
Stepping:                           0x1
BogoMIPS:                           38.40
Flags:                              fp asimd aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm jscvt fcma lrcpc dcpop sha3 sm3 sm4 asimddp sha512 asimdfhm uscat ilrcpc flagm ssbs sb dcpodp flagm2 frint i8mm bf16 rng afp rpres
L1d cache:                          1.1 MiB (12 instances)
L1i cache:                          2.3 MiB (12 instances)
L2 cache:                           144 MiB (12 instances)
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; __user pointer sanitization
Vulnerability Spectre v2:           Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] mypy-boto3-s3==1.36.21
[pip3] numpy==2.2.6
[pip3] numpydoc==1.9.0
[pip3] nvidia-cublas==13.1.0.3
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.19.0.56
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.0
[pip3] nvidia-nccl-cu13==2.28.9
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvtx==13.0.85
[pip3] onnx==1.21.0
[pip3] onnx-ir==0.2.1
[pip3] onnxruntime==1.23.2
[pip3] onnxruntime_extensions==0.15.2
[pip3] onnxscript==0.7.0
[pip3] onnxsim==0.4.36
[pip3] pytorch-ignite==0.5.4
[pip3] pytorch_tokenizers==1.2.0
[pip3] torch==2.14.0+cpu
[pip3] torchao==0.17.0
[pip3] torchvision==0.28.0+cpu
[pip3] triton==3.6.0
[conda] Could not collect

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93

**Comment by abhinav-phi** [comment]
Hi! I'd like to work on this.

Plan: I'll investigate the reported behavior, implement the change/fix per the issue description, and add tests to cover it.

Could you please assign this issue to me? Thanks!

/assign take

**Comment by soulitzer** [comment]
Thanks for the issue, this is an expected bc-breaking change in order to make it consistent with max, etc, if you prefer the alternative behavior, you should write an autograd Function that has the desired backward.
